File: iOS/Mobile/CourseGradesViewController.m
//
//  CourseGradesViewController.m
//  Mobile
//
//  Created by Jason Hocker on 10/17/12.
//  Copyright (c) 2012-2014 Ellucian. All rights reserved.
//

#import "CourseGradesViewController.h"
#import "CurrentUser.h"
#import "AuthenticatedRequest.h"
#import "EmptyTableViewCell.h"
#import "UIViewController+GoogleAnalyticsTrackerSupport.h"

@interface CourseGradesViewController ()
@property (strong, nonatomic) NSDateFormatter *dateFormatter;
@property (strong, nonatomic) NSDateFormatter *datetimeFormatter;
@property (strong, nonatomic) NSDateFormatter *datetimeOutputFormatter;
@end

@implementation CourseGradesViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationController.navigationBar.translucent = NO;
    
    NSError *error;
	if (![[self fetchedResultsController] performFetch:&error]) {
		// Update to handle the error appropriately.
		NSLog(@"Unresolved error %@, %@", error, [error userInfo]);
	}
    
    self.navigationItem.title = self.courseNameAndSectionNumber;
    if([CurrentUser sharedInstance].isLoggedIn) {
        [self fetchGrades:self];
    }
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(fetchGrades:) name:kLoginExecutorSuccess object:nil];
    
    
}

-(void) viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [self sendView:@"Course grades" forModuleNamed:self.module.name];
}

- (NSFetchedResultsController *)fetchedResultsController {
    
    if (_fetchedResultsController != nil) {
        return _fetchedResultsController;
    }
    
    NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] init];
    NSEntityDescription *entity = [NSEntityDescription
                                   entityForName:@"Grade" inManagedObjectContext:self.module.managedObjectContext];
    [fetchRequest setEntity:entity];
    fetchRequest.predicate = [NSPredicate predicateWithFormat:@"course.sectionId == %@ and course.term.termId == %@", self.sectionId, self.termId];
    NSArray *sortDescriptors = [NSArray arrayWithObject:[[NSSortDescriptor alloc]
                                                         initWithKey:@"lastUpdated" ascending:NO]];
    [fetchRequest setSortDescriptors:sortDescriptors];
    
    NSFetchedResultsController *theFetchedResultsController =
    [[NSFetchedResultsController alloc] initWithFetchRequest:fetchRequest
                                        managedObjectContext:self.module.managedObjectContext sectionNameKeyPath:nil
                                                   cacheName:nil];
    self.fetchedResultsController = theFetchedResultsController;
    _fetchedResultsController.delegate = self;
    
    return _fetchedResultsController;
    
}

- (IBAction)dismiss:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    id  sectionInfo = [[_fetchedResultsController sections] objectAtIndex:section];
    NSInteger numberOfObjects = [sectionInfo numberOfObjects];
    
    if(numberOfObjects == 0 ){
        [self showNoDataView:NSLocalizedString(@"No Grades Recorded", @"no grades recorded message")];
    } else {
        [self hideNoDataView];
    }
    return numberOfObjects;
}


#pragma mark - Table view delegate

- (UITableViewCell *)configureAtIndexPath:(NSIndexPath *)indexPath {
   
    Grade *grade = [_fetchedResultsController objectAtIndexPath:indexPath];
                
    GradesCell *cell = [self.tableView dequeueReusableCellWithIdentifier:@"Grades Cell"];
    if (cell == nil) {
                    
        NSArray* views = [[NSBundle mainBundle] loadNibNamed:@"GradesTableViewCell" owner:nil options:nil];
        
        for (UIView *view in views) {
            if([view isKindOfClass:[UITableViewCell class]])
            {
                cell = (GradesCell*)view;
            }
        }
    }
                
    cell.gradeTypeLabel.text = grade.name;
    cell.gradeValueLabel.text = grade.value;
    NSString *formattedDate = [self.datetimeOutputFormatter stringFromDate:grade.lastUpdated];
    if(formattedDate) {
        cell.gradeLastUpdated.text = [NSString stringWithFormat:NSLocalizedString(@"Last Updated: %@", @"label for when data was last updated"), formattedDate];
    }
    
    return cell;
  
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    return [self configureAtIndexPath:indexPath];

}

- (void)controllerWillChangeContent:(NSFetchedResultsController *)controller {
    [self.tableView beginUpdates];
}


- (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)anObject atIndexPath:(NSIndexPath *)indexPath forChangeType:(NSFetchedResultsChangeType)type newIndexPath:(NSIndexPath *)newIndexPath {
    
    UITableView *tableView = self.tableView;
    
    switch(type) {
            
        case NSFetchedResultsChangeInsert:
            [tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath] withRowAnimation:UITableViewRowAnimationFade];
            break;
            
        case NSFetchedResultsChangeDelete:
            [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] withRowAnimation:UITableViewRowAnimationFade];
            break;
            
        case NSFetchedResultsChangeUpdate:
            [self configureAtIndexPath:indexPath];
            break;
            
        case NSFetchedResultsChangeMove:
            [tableView deleteRowsAtIndexPaths:[NSArray
                                               arrayWithObject:indexPath] withRowAnimation:UITableViewRowAnimationFade];
            [tableView insertRowsAtIndexPaths:[NSArray
                                               arrayWithObject:newIndexPath] withRowAnimation:UITableViewRowAnimationFade];
            break;
    }
}


- (void)controller:(NSFetchedResultsController *)controller didChangeSection:(id )sectionInfo atIndex:(NSUInteger)sectionIndex forChangeType:(NSFetchedResultsChangeType)type {
    
    switch(type) {
            
        case NSFetchedResultsChangeInsert:
            [self.tableView insertSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:UITableViewRowAnimationFade];
            break;
            
        case NSFetchedResultsChangeDelete:
            [self.tableView deleteSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:UITableViewRowAnimationFade];
            break;
    }
}


- (void)controllerDidChangeContent:(NSFetchedResultsController *)controller {
    [self.tableView endUpdates];
}

#pragma mark - fetch notifications

- (void) fetchGrades:(id) sender
{
    NSManagedObjectContext *importContext = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSPrivateQueueConcurrencyType];
    importContext.parentContext = self.module.managedObjectContext;
    
    NSString *urlString = [NSString stringWithFormat:@"%@/%@?term=%@&section=%@", [self.module propertyForKey:@"grades"], [[[CurrentUser sharedInstance] userid] stringByAddingPercentEscapesUsingEncoding: NSUTF8StringEncoding], [self.termId stringByAddingPercentEscapesUsingEncoding: NSUTF8StringEncoding] ,[self.sectionId stringByAddingPercentEscapesUsingEncoding: NSUTF8StringEncoding]];
    [importContext performBlock: ^{
        
        //download data
        NSError *error;
        [UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
        AuthenticatedRequest *authenticatedRequet = [AuthenticatedRequest new];
        NSData *responseData = [authenticatedRequet requestURL:[NSURL URLWithString:urlString] fromView:self];
        
        [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
        if(responseData) {
            NSDictionary* json = [NSJSONSerialization
                             JSONObjectWithData:responseData
                             options:kNilOptions
                             error:&error];
            
            NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"GradeTerm"];
            request.predicate = [NSPredicate predicateWithFormat:@"termId == %@", self.termId];
            NSArray *terms = [importContext executeFetchRequest:request error:&error];
            
            //create/update objects
            for(NSDictionary *termJson in [json objectForKey:@"terms"]) {
                GradeTerm *gradeTerm = nil;
                if([self.termId isEqualToString:[termJson objectForKey:@"id"]]) {
                     gradeTerm = [[terms filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"termId == %@", self.termId]] lastObject];
                    if(!gradeTerm) {
                        gradeTerm = [NSEntityDescription insertNewObjectForEntityForName:@"GradeTerm"    inManagedObjectContext:importContext];
                        gradeTerm.termId = [termJson objectForKey:@"id"];
                        gradeTerm.name = [termJson objectForKey:@"name"];
                        gradeTerm.startDate = [self.dateFormatter dateFromString:[termJson objectForKey:@"startDate"]];
                        gradeTerm.endDate = [self.dateFormatter dateFromString:[termJson objectForKey:@"endDate"]];
                    }
                }
                
                for(NSDictionary *courseJson in [termJson objectForKey:@"sections"]) {
                    NSArray* filteredTerms = [terms filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"termId = %@", self.termId]];
                    NSArray *courses = [[[filteredTerms lastObject] courses] array];
                    NSArray *filteredCourses = [courses filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"sectionId = %@", self.sectionId]];
                    GradeCourse *gradeCourse = [filteredCourses lastObject];
                    if(!gradeCourse) {
                    
                        gradeCourse = [NSEntityDescription insertNewObjectForEntityForName:@"GradeCourse" inManagedObjectContext:importContext];
                        gradeCourse.sectionId = [courseJson objectForKey:@"sectionId"];
                        gradeCourse.courseName = [courseJson objectForKey:@"courseName"];
                        gradeCourse.sectionTitle = [courseJson objectForKey:@"sectionTitle"];
                        gradeCourse.courseSectionNumber = [courseJson objectForKey:@"courseSectionNumber"];
                        gradeCourse.term = gradeTerm;
                        [gradeTerm addCoursesObject:gradeCourse];
                    }
                    
                    for (Grade* oldObject in gradeCourse.grades) {
                        [importContext deleteObject:oldObject];
                    }
                    
                    NSArray *gradesJson = [courseJson objectForKey:@"grades"];
                    for(NSDictionary *gradeJson in gradesJson) {
                        
                        Grade *grade = [NSEntityDescription insertNewObjectForEntityForName:@"Grade" inManagedObjectContext:importContext];
                        grade.name = [gradeJson objectForKey:@"name"];
                        grade.value = [gradeJson objectForKey:@"value"];
                        if([gradeJson objectForKey:@"updated"] != [NSNull null]) {
                            grade.lastUpdated = [self.datetimeFormatter dateFromString:[gradeJson objectForKey:@"updated"]];
                        }
                        [gradeCourse addGradesObject:grade];
                        grade.course = gradeCourse;
                    }
                }
            }
            
            if (![importContext save:&error]) {
                NSLog(@"Could not save to main context after update to grades: %@", [error userInfo]);
            }
            
            //persist to store and update fetched result controllers
            [importContext.parentContext performBlock:^{
                NSError *parentError = nil;
                if(![importContext.parentContext save:&parentError]) {
                    NSLog(@"Could not save to store after update to grades: %@", [error userInfo]);
                }
            }];
        }
    }];
    
}

- (NSDateFormatter *)dateFormatter {
    if (_dateFormatter == nil) {
        _dateFormatter = [[NSDateFormatter alloc] init];
        [_dateFormatter setDateFormat:@"yyyy-MM-dd"];
    }
    return _dateFormatter;
}

- (NSDateFormatter *)datetimeFormatter {
    if (_datetimeFormatter == nil) {
        _datetimeFormatter = [[NSDateFormatter alloc] init];
        [_datetimeFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss'Z'"];
        [_datetimeFormatter setTimeZone:[NSTimeZone timeZoneWithName:@"UTC"]];
    }
    return _datetimeFormatter;
}

- (NSDateFormatter *)datetimeOutputFormatter {
    if (_datetimeOutputFormatter == nil) {
        _datetimeOutputFormatter = [[NSDateFormatter alloc] init];
        [_datetimeOutputFormatter setDateStyle:NSDateFormatterShortStyle];
        [_datetimeOutputFormatter setTimeStyle:NSDateFormatterShortStyle];
    }
    return _datetimeOutputFormatter;
}
@end

File: iOS/Mobile/CourseGradesViewController.swift
//
//  CourseGradesViewController.swift
//  Mobile
//
//  Created by Jason Hocker on 10/29/15.
//  Copyright © 2015 Ellucian Company L.P. and its affiliates. All rights reserved.
//

import Foundation

class CourseGradesViewController : UITableViewController, NSFetchedResultsControllerDelegate {
    var termId : String?
    var sectionId : String?
    var courseNameAndSectionNumber : String?
    var module : Module?
    var _fetchedResultsController : NSFetchedResultsController?
    
    let dateFormatter : NSDateFormatter = {
        var formatter = NSDateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()
    let datetimeFormatter : NSDateFormatter = {
        var formatter = NSDateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
        formatter.timeZone = NSTimeZone(name:"UTC")
        return formatter
    }()
    let datetimeOutputFormatter : NSDateFormatter = {
        var formatter = NSDateFormatter()
        formatter.dateStyle = .ShortStyle
        formatter.timeStyle = .ShortStyle
        return formatter
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.navigationController!.navigationBar.translucent = false
        try! self.fetchedResultsController.performFetch()
        self.navigationItem.title = self.courseNameAndSectionNumber
        if CurrentUser.sharedInstance().isLoggedIn {
            self.fetchGrades(self)
        }
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "fetchGrades:", name: kLoginExecutorSuccess, object: nil)
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        self.sendView("Course grades", forModuleNamed: self.module!.name)
    }
    
    var fetchedResultsController: NSFetchedResultsController {
        // return if already initialized
        if self._fetchedResultsController != nil {
            return self._fetchedResultsController!
        }
        let managedObjectContext = self.module!.managedObjectContext!
        
        let request = NSFetchRequest()
        request.entity = NSEntityDescription.entityForName("Grade", inManagedObjectContext: managedObjectContext)
        request.predicate = NSPredicate(format: "course.sectionId == %@ and course.term.termId == %@", self.sectionId!, self.termId!)
        request.sortDescriptors = [NSSortDescriptor(key: "lastUpdated", ascending: false)]
        
        let aFetchedResultsController = NSFetchedResultsController(fetchRequest: request, managedObjectContext: managedObjectContext, sectionNameKeyPath: nil, cacheName: nil)
        aFetchedResultsController.delegate = self
        self._fetchedResultsController = aFetchedResultsController
        
        do {
            try self._fetchedResultsController!.performFetch()
            
        } catch let error {
            print("fetch error: \(error)")
        }
        
        return self._fetchedResultsController!
    }
    
    @IBAction func dismiss(sender: AnyObject) {
        self.dismissViewControllerAnimated(true, completion: nil)
    }
    
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if let sections = fetchedResultsController.sections {
            let currentSection = sections[section] as NSFetchedResultsSectionInfo
            let count = currentSection.numberOfObjects
            if count > 0 {
                return count
            } else {
                return 1
            }
        }
        
        return 1
    }
    
    
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        
        if fetchedResultsController.fetchedObjects?.count > 0 {
            
            let grade = fetchedResultsController.objectAtIndexPath(indexPath)
            let cell = tableView.dequeueReusableCellWithIdentifier("Grade Cell", forIndexPath: indexPath) as UITableViewCell
            
            
            let descriptionLabel = cell.viewWithTag(1) as! UILabel
            let lastUpdatedLabel = cell.viewWithTag(2) as! UILabel
            let gradeLabel = cell.viewWithTag(3) as! UILabel
            
            descriptionLabel.text = grade.name
            
            if let date = grade.lastUpdated where date != nil {
                let formattedDate = self.datetimeOutputFormatter.stringFromDate(date)
                lastUpdatedLabel.text = String(format: NSLocalizedString("Last Updated %@", comment: "Last Updated date"), formattedDate)
            } else {
                lastUpdatedLabel.text = NSLocalizedString("Last Updated Unknown", comment: "Last Updated date unknown")
            }
            
            gradeLabel.text = grade.value
            return cell
        } else {
            return tableView.dequeueReusableCellWithIdentifier("No Grades Cell", forIndexPath: indexPath) as UITableViewCell
        }
    }

    
    func fetchGrades(sender: AnyObject) {
        
        
        if let userid = CurrentUser.sharedInstance().userid {
            let urlBase = self.module?.propertyForKey("grades")
            let escapedUserId = userid.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.URLQueryAllowedCharacterSet())
            let escapedTermId = self.termId?.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.URLQueryAllowedCharacterSet())
            let escapedSectionId = self.sectionId?.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.URLQueryAllowedCharacterSet())
            let urlString = "\(urlBase!)/\(escapedUserId!)?term=\(escapedTermId!)&section=\(escapedSectionId!)"
            
            if self.fetchedResultsController.fetchedObjects!.count <= 0 {
                let hud = MBProgressHUD.showHUDAddedTo(self.view, animated: true)
                hud.labelText = NSLocalizedString("Loading", comment: "loading message while waiting for data to load")
            }
            
            let privateContext = NSManagedObjectContext(concurrencyType: .PrivateQueueConcurrencyType)
            privateContext.parentContext = self.module!.managedObjectContext
            privateContext.undoManager = nil
            
            privateContext.performBlock { () -> Void in
                
                do {
                    UIApplication.sharedApplication().networkActivityIndicatorVisible = true
                    
                    defer {
                        dispatch_async(dispatch_get_main_queue()) {
                            
                            UIApplication.sharedApplication().networkActivityIndicatorVisible = false
                            MBProgressHUD.hideHUDForView(self.view, animated: true)
                        }
                    }
                    
                    let authenticatedRequest = AuthenticatedRequest()
                    let responseData = authenticatedRequest.requestURL(NSURL(string: urlString)!, fromView: self)
                    let json = JSON(data: responseData)
                    UIApplication.sharedApplication().networkActivityIndicatorVisible = false
                    
                    let request = NSFetchRequest(entityName: "GradeTerm")
                    request.predicate = NSPredicate(format: "termId == %@", self.termId!)
                    let terms = try privateContext
                        .executeFetchRequest(request) as! [GradeTerm]
                    
                    if let jsonTerms = json["terms"].array {
                        for jsonTerm in jsonTerms {
                            
                            
                            
                            let filtered = terms.filter {
                                let x = $0 as GradeTerm
                                return x.termId == self.termId
                            }
                            let gradeTerm : GradeTerm
                            if filtered.count > 0 {
                                gradeTerm = filtered.first!
                            } else {
                                gradeTerm = NSEntityDescription.insertNewObjectForEntityForName("GradeTerm", inManagedObjectContext: privateContext) as! GradeTerm
                                gradeTerm.termId = jsonTerm["id"].string
                                gradeTerm.name = jsonTerm["name"].string
                                gradeTerm.startDate = self.dateFormatter.dateFromString(jsonTerm["startDate"].string!)
                                gradeTerm.endDate = self.dateFormatter.dateFromString(jsonTerm["endDate"].string!)
                                
                                
                            }
                            if let jsonSections = jsonTerm["sections"].array {
                                for jsonSection in jsonSections {
                                    let courses = gradeTerm.courses
                                    let filtered = courses!.filter {
                                        let x = $0 as! GradeCourse
                                        return x.sectionId == self.sectionId
                                    }
                                    let gradeCourse : GradeCourse
                                    if filtered.count > 0 {
                                        gradeCourse = filtered.first as! GradeCourse
                                    } else {
                                        gradeCourse = NSEntityDescription.insertNewObjectForEntityForName("GradeCourse", inManagedObjectContext: privateContext) as! GradeCourse
                                        gradeCourse.sectionId = jsonSection["sectionId"].string
                                        gradeCourse.courseName = jsonSection["courseName"].string
                                        gradeCourse.sectionTitle = jsonSection["sectionTitle"].string
                                        gradeCourse.courseSectionNumber = jsonSection["courseSectionNumber"].string
                                        gradeCourse.term = gradeTerm
                                        gradeTerm.addCoursesObject(gradeCourse)
                                    }
                                    
                                    for oldObject in gradeCourse.grades {
                                        privateContext.deleteObject(oldObject as! Grade)
                                    }
                                    
                                    if let jsonGrades = jsonSection["grades"].array {
                                        for jsonGrade in jsonGrades {
                                            let grade = NSEntityDescription.insertNewObjectForEntityForName("Grade", inManagedObjectContext: privateContext) as! Grade
                                            grade.name = jsonGrade["name"].string
                                            grade.value = jsonGrade["value"].string
                                            if let lastUpdated = jsonGrade["updated"].string {
                                                grade.lastUpdated = self.datetimeFormatter.dateFromString(lastUpdated)
                                            }
                                            gradeCourse.addGradesObject(grade)
                                            grade.course = gradeCourse
                                            
                                        }
                                    }
                                    
                                }
                            }
                        }
                    }
                    try privateContext.save()
                    
                    privateContext.parentContext?.performBlock({
                        do {
                            try privateContext.parentContext?.save()
                        } catch let error {
                            print (error)
                        }
                    })
                    
                    dispatch_async(dispatch_get_main_queue(), {
                        self._fetchedResultsController = nil;
                        try! self.fetchedResultsController.performFetch()
                        self.tableView.reloadData()
                    })
                    
                } catch let error {
                    print (error)
                }
                
            }
        }
        
    }
    
    
}